File: RB110/ruby_interview_problems/new_rb119_interview_study_problems/problem2.rb
=begin
Create a method that takes an array of integers as an argument. The method should 
return the minimum sum of 5 consecutive numbers in the array. If the array contains 
fewer than 5 elements, the method should return nil.

Examples:

p minimum_sum([1, 2, 3, 4]) == nil
p minimum_sum([1, 2, 3, 4, 5, -5]) == 9
p minimum_sum([1, 2, 3, 4, 5, 6]) == 15
p minimum_sum([55, 2, 6, 5, 1, 2, 9, 3, 5, 100]) == 16
p minimum_sum([-1, -5, -3, 0, -1, 2, -4]) == -10

Problem:
- Return the minimum sum of 5 CONSECUTIVE numbers from the given array
- If the given array has less than 5 integers, return nil
- In this case, CONSECUTIVE means one after the other
- numbers in array may be negative as well

- Input: array of integers
- Output: minimal sum integer

Examples:
[1, 2, 3, 4] -> nil
  - We reutrn nil bc side is less than 5

[1, 2, 3, 4, 5, -5] -> 9
  - [1, 2, 3, 4, 5] == 15 or [2, 3, 4, 5, -5] == 9
    - return 9 as it is the smallest sum

Data Structures:
- maintaing array would easily allow to come up with consecutive array
  combinations as well as sum each array easily

Algorithm:
- RETURN nil if array size is less than 5
- CREATE a results array
- CREATE all possible subarrays of 5 consecutive integers
  - FIND its sum and append it to the results array
- RETURN the smallest num in the results array
=end

def minimum_sum(numbers)
  return nil if numbers.size < 5
  all_sums = []

  (numbers.size - 4).times do |start_point|
    all_sums << numbers[start_point..start_point + 4].sum
  end
  
  all_sums.min
end

p minimum_sum([1, 2, 3, 4]) == nil
p minimum_sum([1, 2, 3, 4, 5, -5]) == 9
p minimum_sum([1, 2, 3, 4, 5, 6]) == 15
p minimum_sum([55, 2, 6, 5, 1, 2, 9, 3, 5, 100]) == 16
p minimum_sum([-1, -5, -3, 0, -1, 2, -4]) == -10

File: LS215/JS_Small_Problems/LS215_Exam_Practice_Problems/problem2.js
// Write a function that displays a four-pointed diamond in an nxn grid, where n is an odd integer supplied as an argument to the function. You may assume that the argument will always be an odd integer.

/*
  ----- PROBLEM -----
INPUT: Given an integer
OUTPUT: 4 pointed `diamond` logged to console

Rules:
  - Input will always be an odd integer
  - Logged `diamond` will print '*' to the console in an `n` by `n` grid,
    where `n` is the odd integer supplied
  - If `1` oassed to func, return `*`
  - Number passed to argument correlates with lines of stars logged to console
  - Top half is `reflected` in bottom half
  - Mid point of `diamond` contains `n` stars
  - Each row of stars above and below the mid point change by +/- 2 stars
  - If no stars, fill up grid with `' '`
  - Midline can be determined by `n` / 2 and rounding up
    - 5 / 2 = 2.5 -> 3
    - 9 / 2 = 4.5 -> 5
Assumptions:
  - If integer passed is negative
    - Return null
  - If non odd integer is passed
    - Return null
  - Is there a min or max limit on integer argument?
    - Assume no max limit, min limit will be 1

  ----- EXAMPLES -----
diamond(3);
  // logs
 *  | whitespace star whitespace - 1 star, + 2stars next line
*** | star star star            -> midline, 3 stars, - 2 stars next line
 *  | whitespace star whitespace -> 1 star

diamond(5);
  // logs
  *  |  top of diamond, 1 star
 *** |  top of diamond 3 stars
*****| star star star star star -> midline 5 stars
 *** |  bottom of diamond, 3 stars
  *  |  bottom of diamond 1 star

  ----- DATA STRUCTURES -----
- Array would allow easy iteration, but do not need an ordered list of items
- Object would provide for paired objects, but no need for associating any Key / Values...
- Use for loop to create diamond

  ----- ALGORITHM -----
- Check if Valid argument
  - Return null if argument is not an odd positive integer
- LOG first half of diamond
  - Find starting space amount and set to space variable
    - Math.floor( `n` / 2);
  - Set stars to 1
  - set linesOfStars to 0
  - log to console '{' ' * multipler}{* times multipler}{' ' * multipler}'
  - if linesofstars is less than midpoint Math.ceil(`n`  / 2)
    - decrement space value by - 1
  - If lineof stars is more than midpoint
    - increment space value by + 1
  - If linesOfStars is less than midpoint
    - Increment stars by + 2
  - If linesof stars is equal to or more than midpoint  
    - decrement stars by - 2
  - break out of loop when linesofstars is <= `n`

*/

function diamond(sizeOfDiamond) {
  if (!Number.isInteger(sizeOfDiamond) ||
      sizeOfDiamond <= 0 ||
      sizeOfDiamond % 2 === 0) return null

  let space = (Math.floor(sizeOfDiamond / 2));
  let stars = 1;
  let midpoint = Math.ceil(sizeOfDiamond / 2);

  for (let currentLine = 1; currentLine <= sizeOfDiamond; currentLine += 1) {
    console.log(`${' '.repeat(space)}${'*'.repeat(stars)}${' '.repeat(space)}`);
    if (currentLine < midpoint) space -= 1;
    if (currentLine >= midpoint) space += 1;
    if (currentLine < midpoint) stars += 2;
    if (currentLine >= midpoint) stars -= 2;
  }
}

// FIGURE OUT HOW TO REPEAT CHARACTERS

diamond(1);
// logs
// *

diamond(3);
// logs
//  *
// ***
//  *

diamond(9);
// logs
//     *
//    ***
//   *****
//  *******
// *********
//  *******
//   *****
//    ***
//     *

console.log(diamond(0)); // null
console.log(diamond('hi')); // null
console.log(diamond(-1)); // null